### PR TITLE
Pub 1073 subs man trigger

### PR DESCRIPTION
**Pull request by furnivall**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PUB-1073

### Change description ###
- Added subs trigger from data-management to subscription-management which checks the display from date and sends a trigger to subs-man if required.
- Uses the webClient library which is apparently the RestTemplate successor.
- Uses MockWebServer for testing.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 1]
```diff
@@ -129,6 +129,11 @@ public PublicationController(PublicationService publicationService, ValidationSe
         Artefact createdItem = publicationService
             .createPublication(artefact, payload);
 
+        //TODO handle the below in some way
+        @SuppressWarnings("PMD.PrematureDeclaration")
+        List<Subscription> unusedListOfEligibleSubscribers =
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:134]: Just thinking through the scenario's. Is the unused list needed in data management? It feels more like a subscription-management thing

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Channel.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.hmcts.reform.pip.data.management.models.external;
+
+public enum Channel {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Channel.java:3]: Don't believe this should be in data-management. Responsibility is more on sub / channel management

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/SearchType.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.pip.data.management.models.external;
+
+public enum SearchType {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/SearchType.java:3]: Same comment as above

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Subscription.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package uk.gov.hmcts.reform.pip.data.management.models.external;
+
+
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+@NoArgsConstructor
+public class Subscription {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Subscription.java:10]: Same comment as above

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 5]
```diff
@@ -68,10 +74,22 @@ public Artefact createPublication(Artefact artefact, MultipartFile file) {
...
     }
 
+
+    /**
+     * Checks if the artefact has a display from date of today or previous then triggers the sub fulfilment
+     * process on subscription-management if appropriate.
+     */
+    public List<Subscription> checkAndTriggerSubscriptionManagement(Artefact artefact) {
+        if (artefact.getDisplayFrom().isBefore(LocalDateTime.now().plusDays(1))) {
+            return subscriptionManagementService.sendSubTrigger(artefact);
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:87]: I think the logic for the date checking will need updating a bit, as time has the chance to be included as well in dates so not sure the plus one days will work.

Also I would check the displayTo instead, just to make sure it's in the range.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 6]
```diff
@@ -68,10 +74,22 @@ public Artefact createPublication(Artefact artefact, MultipartFile file) {
...
+        } else {
+            return Collections.emptyList();
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:89]: Don't believe we need the triggered subs in data management.

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementService.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+@Slf4j
+@Component
+public class SubscriptionManagementService {
+
+    @Value("${service-to-service.subscription-management}")
+    private final String url;
+
+    @Autowired
+    public SubscriptionManagementService(@Value("${service-to-service.subscription-management}") String url) {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementService.java:27]: Not sure you need this @Autowired :) 

The @Value should be enough to cover this

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java (hunk 1 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java]: sorted.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Channel.java (hunk 2 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Channel.java]: sorted.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/SearchType.java (hunk 3 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/SearchType.java]: sorted

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Subscription.java (hunk 4 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/models/external/Subscription.java]: sorted

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 5 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:87]: Would be good if you could take another look at this. Seems to work in all cases, including that of a general_publication with no date from/to published

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 6 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java]: sorted

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementService.java (hunk 7 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementService.java:21]: cheers, yep sorted.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 8]
```diff
@@ -205,10 +230,4 @@ public void checkNewlyActiveArtefacts() {
         List<Artefact> newArtefactsToday = artefactRepository.findArtefactsByDisplayFrom(LocalDate.now());
         newArtefactsToday.forEach(artefact -> log.info(sendArtefactForSubscription(artefact)));
     }
-
-    public String sendArtefactForSubscription(Artefact artefact) {
-        //TODO: PUB 1001 check which subscribers are valid here
-        return subscriptionManagementService.getSubscribersToArtefact(artefact);
-    }
-
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:213]: ???

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 9]
```diff
@@ -76,10 +78,29 @@ public Artefact createPublication(Artefact artefact, MultipartFile file) {
...
+            log.info(sendArtefactForSubscription(artefact));
+        } else {
+            log.error("invalid publication, no trigger sent");
+        }
+    }
+
+    public String sendArtefactForSubscription(Artefact artefact) {
+        //TODO: PUB 1001 check which subscribers are valid here
+        return subscriptionManagementService.sendArtefactForSubscription(artefact);
+    }
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:102]: oh wait, youve moved it??? whats the reasoning for hiding this new logic in the middle of the service file?

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java [hunk 10]
```diff
@@ -1,89 +1,53 @@
...
-        HttpServerErrorException httpServerErrorException =
-            new HttpServerErrorException(HttpStatus.BAD_GATEWAY, "Bad Gateway", null, null);
-        doThrow(httpServerErrorException).when(restTemplate)
-            .postForEntity(SUBSCRIPTION_URL, artefact, String.class);
-
-        assertEquals("Subscription trigger unsuccessful for artefact: " + ARTEFACT_ID,
-                     subscriptionManagementService.getSubscribersToArtefact(artefact),
-                     SHOULD_MATCH);
+    void testSendTrigger() throws IOException {
+        mockSubscriptionManagementEndpoint = new MockWebServer();
```
Comment (kurt-uk) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java:39]: the docs are saying this is a integration test?

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 9 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:102]: I just put it directly underneath the trigger method, wasn't trying to cause issues.

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java [hunk 11]
```diff
@@ -1,89 +1,53 @@
...
-        HttpClientErrorException httpClientErrorException =
-            new HttpClientErrorException(HttpStatus.BAD_GATEWAY, "Bad Gateway", null, null);
-        doThrow(httpClientErrorException).when(restTemplate)
-            .postForEntity(SUBSCRIPTION_URL, artefact, String.class);
-        assertEquals("Subscription trigger unsuccessful for artefact: " + ARTEFACT_ID,
-                     subscriptionManagementService.getSubscribersToArtefact(artefact),
-                     SHOULD_MATCH);
+    void testFailedSend() {
+        assertEquals(subscriptionManagementService.sendArtefactForSubscription(ARTEFACT), "Request failed",
+                     "Trigger failed to send.");
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java:51]: So - about this - the WebClientException is always going to be the one that's going to be thrown as the URI is essentially hardcoded, but calling new URI requires there to be either a catch here or adding to the method sig of every call. What's the right approach @kurt-uk?

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java (hunk 10 above)
Comment (furnivall) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java:39]: https://www.baeldung.com/spring-mocking-webclient I went through this and it heavily recommended using mockwebserver vs traditional mocks (i.e. mockito)

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java [hunk 12]
```diff
@@ -1,89 +1,53 @@
...
-import java.util.UUID;
-import java.util.concurrent.ConcurrentHashMap;
+import java.io.IOException;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.mockito.Mockito.doThrow;
-import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
+@ExtendWith(OutputCaptureExtension.class)
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java:21]: Cheers - fixed

File: src/main/resources/application.yaml [hunk 13]
```diff
@@ -56,6 +56,9 @@ azure:
     connection-string: ${CONNECTION_STRING}
     container-name: artefact
 
+service-to-service:
+  subscription-management: ${SUBSCRIPTION_MANAGEMENT_URL:https://pip-subscription-management.staging.platform.hmcts.net/}
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/resources/application.yaml:60]: Aha - cheers

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 14]
```diff
@@ -197,6 +204,25 @@ public Resource getFlatFileByArtefactID(UUID artefactId, Boolean verification) {
...
+     * process on subscription-management if appropriate.
+     */
+    public void checkAndTriggerSubscriptionManagement(Artefact artefact) {
+        //TODO: fully switch this logic to localdates once artefact model changes
+        if (artefact.getDisplayFrom().toLocalDate().isBefore(LocalDate.now().plusDays(1))
+            && (artefact.getDisplayTo() == null
+            || artefact.getDisplayTo().toLocalDate().isAfter(LocalDate.now().minusDays(1)))) {
+            log.info(sendArtefactForSubscription(artefact));
+        } else {
+            log.error("invalid publication, no trigger sent");
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:218]: I'm not sure we want to log an error here, as a publication in the future wouldn't be an error scenario (just picked up later)

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 14 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java]: Have now switched it to an info log - is that okay?

Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java]: do we even need to log this non action? 🤔

Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java]: Can just chuck it then! No worries :)

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java [hunk 15]
```diff
@@ -220,10 +256,12 @@ void testArtefactContentFromAzureWhenUnauthorized() {
...
-        assertThrows(NotFoundException.class, ()
-            -> publicationService.getPayloadByArtefactId(ARTEFACT_ID, true),
-                      "Not Found exception has not been thrown when artefact does not exist");
-
+        assertThrows(
+            NotFoundException.class,
+            ()
+                -> publicationService.getPayloadByArtefactId(ARTEFACT_ID, true),
+            "Not Found exception has not been thrown when artefact does not exist"
+        );
```
Comment (kurt-uk) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java:264]: whats with all the crazy formatting you do to all the files 😂

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java [hunk 16]
```diff
@@ -417,9 +464,46 @@ void testInvalidEnumTypeThrows() {
...
+        when(subscriptionManagementService.sendArtefactForSubscription(artefactFromThePast)).thenReturn(
+            SUCCESSFUL_TRIGGER);
+        try (LogCaptor logCaptor = LogCaptor.forClass(PublicationService.class)) {
+            publicationService.checkAndTriggerSubscriptionManagement(artefactFromThePast);
+            assertEquals(SUCCESSFUL_TRIGGER, logCaptor.getInfoLogs().get(0),
+                         "Should have returned the subscription list"
+            );
+        }
+    }
+
```
Comment (kurt-uk) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java:503]: looking at the O notation for the logic in pub services, you should have at least 5 tests for your date logic, some obvious missing ones i can see is the null values, and if the display from and to are the same date

**Review by kurt-uk — CHANGES_REQUESTED** [review]

**Review by ChrisS1512 — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java (hunk 16 above)
Comment (furnivall) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java:525]: Potentially sorted....

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java [hunk 17]
```diff
@@ -1,89 +1,63 @@
...
 
-@ExtendWith(MockitoExtension.class)
+@ExtendWith({MockitoExtension.class})
 @ActiveProfiles("test")
-@SpringBootTest(classes = {Application.class, AzureBlobConfigurationTest.class, RestTemplateConfigurationTest.class})
+@SpringBootTest(classes = {Application.class})
 @AutoConfigureEmbeddedDatabase(type = AutoConfigureEmbeddedDatabase.DatabaseType.POSTGRES)
 class SubscriptionManagementServiceTest {
+    @Value("${service-to-service.subscription-management}")
+    String url;
```
Comment (kurt-uk) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/SubscriptionManagementServiceTest.java:28]: this doesnt look like its used?

**Review by kurt-uk — CHANGES_REQUESTED** [review]

**Review by kurt-uk — APPROVED** [review]

**Review by kurt-uk — APPROVED** [review]

**Review by ChrisS1512 — APPROVED** [review]

File: build.gradle [hunk 18]
```diff
@@ -186,7 +187,11 @@ dependencies {
   implementation 'com.vladmihalcea:hibernate-types-52:2.14.0'
   implementation 'org.apache.tomcat.embed:tomcat-embed-core:9.0.58'
   implementation 'org.apache.tomcat.embed:tomcat-embed-websocket:9.0.58'
+  implementation 'org.springframework.boot:spring-boot-starter-webflux'
+  implementation("com.squareup.okhttp3:okhttp:4.9.3")
+  testImplementation("com.squareup.okhttp3:mockwebserver:4.9.3")
   testImplementation "io.zonky.test:embedded-database-spring-test:2.1.1"
+  testImplementation group: 'io.github.hakky54', name: 'logcaptor', version: '2.7.8'
```
Comment (kurt-uk) [inline build.gradle:194]: this is duplicated

Comment (furnivall) [reply build.gradle]: fixed now :)

**Pull request merged**